internal/utils/crossnamespace/kongreferencegrant.go

```diff
@@ -0,0 +1,172 @@
+package crossnamespace
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	configurationv1alpha1 "github.com/kong/kong-operator/api/configuration/v1alpha1"
+)
+
+// ErrCrossNamespaceReferenceNotGranted is an error type that indicates a cross-namespace
+// reference is not granted by any KongReferenceGrant.
+type ErrCrossNamespaceReferenceNotGranted struct {
+	FromNamespace string
+	FromGVK       metav1.GroupVersionKind
+	ToNamespace   string
+	ToName        string
+	ToGVK         metav1.GroupVersionKind
+}
+
+// Error returns a formatted error message indicating that a cross-namespace reference
+// is not permitted.
+func (e *ErrCrossNamespaceReferenceNotGranted) Error() string {
+	return fmt.Sprintf("cross-namespace reference to %s %s in namespace %s from %s.%s in namespace %s is not granted by any KongReferenceGrant",
+		e.ToGVK.Kind,
+		e.ToName,
+		e.ToNamespace,
+		e.FromGVK.Kind,
+		e.FromGVK.Group,
+		e.FromNamespace,
+	)
+}
+
+// IsCrossNamespaceReferenceNotGranted checks if the provided error is or wraps
+// an ErrCrossNamespaceReferenceNotGranted error, indicating that a cross-namespace
+// reference was attempted without the proper ReferenceGrant permissions.
+// It returns true if the error matches this type, false otherwise.
+func IsCrossNamespaceReferenceNotGranted(err error) bool {
```

Comment:
Since we have the package which is called `crossnamespace`, I'd suggest to simplify the name here to something like: 

```suggestion
// IsReferenceNotGranted checks if the provided error is or wraps
// an ErrCrossNamespaceReferenceNotGranted error, indicating that a cross-namespace
// reference was attempted without the proper ReferenceGrant permissions.
// It returns true if the error matches this type, false otherwise.
func IsNotGranted(err error) bool {
```

This will prevent the stutter in package + func name.